rotkehlchen/chain/ethereum/modules/curve/decoder.py

```diff
@@ -283,13 +294,22 @@ def _decode_curve_deposit_events(
                 event.counterparty = CPT_CURVE
                 event.notes = f'Deposit {event.balance.amount} {crypto_asset.symbol} in curve pool {tx_log.address}'  # noqa: E501
                 deposit_events.append(event)
+            elif (
+                event.event_type == HistoryEventType.DEPOSIT and
+                event.event_subtype == HistoryEventSubType.DEPOSIT_ASSET and
+                event.counterparty == CPT_CURVE
+            ):
+                already_processed = True
 
```

Comment:
This is incorrect. What happens is that we iterate multiple times over the decoded events. This happens because we try to decode events every time they match a decoding rule in 

```
decoding_output = self.decode_by_address_rules(tx_log, transaction, events, tx_receipt.logs, action_items)  # noqa: E501
```

So here this function is first called with for example 2 events and then called again with 3 events in`decoded_events`. You can check in `test_3pool_add_liquidity`. 

if you exit early as you are doing then you will miss the `_sort_events` logic that I believe is needed in some cases. You can check it yourself.

I'm not really sure what guided you to add this piece of code here

Reply:
What I'm not sure is why no test failed after your change. I believe as they are there is no deposit + zap or the conditions that require the logic happening after

Reply:
> if you exit early as you are doing then you will miss the _sort_events logic that I believe is needed in some cases. You can check it yourself.

If you check closer, 

```python
if already_processed is True:
                return DEFAULT_DECODING_OUTPUT
```

this condition happens only inside

```python
if receive_event is None or len(deposit_events) == 0:
```

so it does not affect sorting.

-------------

The only thing that it may early-exit before is `return DecodingOutput(matched_counterparty=CPT_CURVE)`, which could disable the trigger of the needed post decoding rule for zaps. But all is fine here since `already_processed` can be `True` only if it is the second call this curve decoding method during processing a transaction, and in this case, if it is a zap call, `return DecodingOutput(matched_counterparty=CPT_CURVE)` will happen during the first call and the zaps logic will be handled properly.

Reply:
I can also move this check

```python
if already_processed is True:
                return DEFAULT_DECODING_OUTPUT
```

after
```python
if user_or_contract_address in CURVE_DEPOSIT_CONTRACTS and len(deposit_events) > 0:
                return DecodingOutput(matched_counterparty=CPT_CURVE)
```

to be 100% sure that the post decoding rule trigger is not missed.